File: app/src/main/java/com/fastaccess/ui/adapter/viewholder/FeedsViewHolder.java
package com.fastaccess.ui.adapter.viewholder;

import android.content.res.Resources;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;

import com.fastaccess.R;
import com.fastaccess.data.dao.GitCommitModel;
import com.fastaccess.data.dao.LabelModel;
import com.fastaccess.data.dao.PayloadModel;
import com.fastaccess.data.dao.TeamsModel;
import com.fastaccess.data.dao.WikiModel;
import com.fastaccess.data.dao.model.Comment;
import com.fastaccess.data.dao.model.Event;
import com.fastaccess.data.dao.model.Issue;
import com.fastaccess.data.dao.model.PullRequest;
import com.fastaccess.data.dao.model.Release;
import com.fastaccess.data.dao.model.User;
import com.fastaccess.data.dao.types.EventsType;
import com.fastaccess.helper.ParseDateFormat;
import com.fastaccess.provider.markdown.MarkDownProvider;
import com.fastaccess.provider.scheme.LinkParserHelper;
import com.fastaccess.ui.widgets.AvatarLayout;
import com.fastaccess.ui.widgets.FontTextView;
import com.fastaccess.ui.widgets.SpannableBuilder;
import com.fastaccess.ui.widgets.recyclerview.BaseRecyclerAdapter;
import com.fastaccess.ui.widgets.recyclerview.BaseViewHolder;

import java.util.List;

import butterknife.BindView;

/**
 * Created by Kosh on 11 Nov 2016, 2:08 PM
 */

public class FeedsViewHolder extends BaseViewHolder<Event> {

    @Nullable @BindView(R.id.avatarLayout) AvatarLayout avatar;
    @BindView(R.id.description) FontTextView description;
    @BindView(R.id.title) FontTextView title;
    @BindView(R.id.date) FontTextView date;
    private Resources resources;

    public FeedsViewHolder(@NonNull View itemView, @Nullable BaseRecyclerAdapter adapter) {
        super(itemView, adapter);
        this.resources = itemView.getResources();
    }

    public static View getView(@NonNull ViewGroup viewGroup, boolean noImage) {
        if (noImage) {
            return getView(viewGroup, R.layout.feeds_row_no_image_item);
        } else {
            return getView(viewGroup, R.layout.feeds_row_item);
        }
    }

    @Override public void bind(@NonNull Event eventsModel) {
        appendAvatar(eventsModel);
        SpannableBuilder spannableBuilder = SpannableBuilder.builder();
        appendActor(eventsModel, spannableBuilder);
        description.setMaxLines(2);
        description.setText("");
        description.setVisibility(View.GONE);
        if (eventsModel.getType() != null) {
            EventsType type = eventsModel.getType();
            if (type == EventsType.WatchEvent) {
                appendWatch(spannableBuilder, type, eventsModel);
            } else if (type == EventsType.CreateEvent) {
                appendCreateEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.CommitCommentEvent) {
                appendCommitComment(spannableBuilder, eventsModel);
            } else if (type == EventsType.DownloadEvent) {
                appendDownloadEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.FollowEvent) {
                appendFollowEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.ForkEvent) {
                appendForkEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.GistEvent) {
                appendGistEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.GollumEvent) {
                appendGollumEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.IssueCommentEvent) {
                appendIssueCommentEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.IssuesEvent) {
                appendIssueEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.MemberEvent) {
                appendMemberEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.PublicEvent) {
                appendPublicEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.PullRequestEvent) {
                appendPullRequestEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.PullRequestReviewCommentEvent) {
                appendPullRequestReviewCommentEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.PullRequestReviewEvent) {
                appendPullRequestReviewCommentEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.RepositoryEvent) {
                appendPublicEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.PushEvent) {
                appendPushEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.TeamAddEvent) {
                appendTeamEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.DeleteEvent) {
                appendDeleteEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.ReleaseEvent) {
                appendReleaseEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.ForkApplyEvent) {
                appendForkApplyEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.OrgBlockEvent) {
                appendOrgBlockEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.ProjectCardEvent) {
                appendProjectCardEvent(spannableBuilder, eventsModel, false);
            } else if (type == EventsType.ProjectColumnEvent) {
                appendProjectCardEvent(spannableBuilder, eventsModel, true);
            } else if (type == EventsType.OrganizationEvent) {
                appendOrganizationEvent(spannableBuilder, eventsModel);
            } else if (type == EventsType.ProjectEvent) {
                appendProjectCardEvent(spannableBuilder, eventsModel, false);
            }
            date.setGravity(Gravity.CENTER);
            date.setEventsIcon(type.getDrawableRes());
        }
        title.setText(spannableBuilder);
        date.setText(ParseDateFormat.getTimeAgo(eventsModel.getCreatedAt()));
    }

    private void appendOrganizationEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        spannableBuilder.bold(eventsModel.getPayload().getAction().replaceAll("_", ""))
                .append(" ")
                .append(eventsModel.getPayload().getInvitation() != null ? eventsModel.getPayload().getInvitation().getLogin() + " " : "")
                .append(eventsModel.getPayload().getOrganization().getLogin());
    }

    private void appendProjectCardEvent(SpannableBuilder spannableBuilder, Event eventsModel, boolean isColumn) {
        spannableBuilder.bold(eventsModel.getPayload().getAction())
                .append(" ")
                .append(!isColumn ? "project" : "column")
                .append(" ")
                .append(eventsModel.getRepo().getName());
    }

    private void appendOrgBlockEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        spannableBuilder.bold(eventsModel.getPayload().getAction())
                .append(" ")
                .append(eventsModel.getPayload().getBlockedUser().getLogin())
                .append(" ")
                .append(eventsModel.getPayload().getOrganization().getLogin());
    }

    private void appendForkApplyEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        spannableBuilder.bold(eventsModel.getPayload().getHead())
                .append(" ")
                .append(eventsModel.getPayload().getBefore())
                .append(" ")
                .append(eventsModel.getRepo() != null ? "in " + eventsModel.getRepo().getName() : "");
    }

    private void appendReleaseEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        Release release = eventsModel.getPayload().getRelease();
        spannableBuilder.bold("released")
                .append(" ")
                .append(release.getName())
                .append(" ")
                .append(eventsModel.getRepo().getName());
    }

    private void appendDeleteEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        spannableBuilder.bold("deleted")
                .append(" ")
                .append(eventsModel.getPayload().getRefType())
                .append(" ")
                .append(eventsModel.getPayload().getRef())
                .append(" ")
                .bold("at")
                .append(" ")
                .append(eventsModel.getRepo().getName());
    }

    private void appendTeamEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        TeamsModel teamsModel = eventsModel.getPayload().getTeam();
        User user = eventsModel.getPayload().getUser();
        spannableBuilder.bold("added")
                .append(" ")
                .append(user != null ? user.getLogin() : eventsModel.getRepo().getName())
                .append(" ")
                .bold("in")
                .append(" ")
                .append(teamsModel.getName() != null ? teamsModel.getName() : teamsModel.getSlug());
    }

    private void appendPushEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        String ref = eventsModel.getPayload().getRef();
        if (ref.startsWith("refs/heads/")) {
            ref = ref.substring(11);
        }
        spannableBuilder.bold("pushed to")
                .append(" ")
                .append(ref)
                .append(" ")
                .bold("at")
                .append(" ")
                .append(eventsModel.getRepo().getName());
        final List<GitCommitModel> commits = eventsModel.getPayload().getCommits();
        int size = commits != null ? commits.size() : -1;
        SpannableBuilder spanCommits = SpannableBuilder.builder();
        if (size > 0) {
            if (size != 1) spanCommits.append(String.valueOf(eventsModel.getPayload().getSize())).append(" new commits").append("\n");
            else spanCommits.append("1 new commit").append("\n");
            int max = 5;
            int appended = 0;
            for (GitCommitModel commit : commits) {
                if (commit == null) continue;
                String sha = commit.getSha();
                if (TextUtils.isEmpty(sha)) continue;
                sha = sha.length() > 7 ? sha.substring(0, 7) : sha;
                spanCommits.url(sha).append(" ")
                        .append(commit.getMessage() != null ? commit.getMessage().replaceAll("\\r?\\n|\\r", " ") : "")
                        .append("\n");
                appended++;
                if (appended == max) break;
            }
        }
        if (spanCommits.length() > 0) {
            int last = spanCommits.length();
            description.setMaxLines(5);
            description.setText(spanCommits.delete(last - 1, last));
            description.setVisibility(View.VISIBLE);
        } else {
            description.setText("");
            description.setMaxLines(2);
            description.setVisibility(View.GONE);
        }
    }

    private void appendPullRequestReviewCommentEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        PullRequest pullRequest = eventsModel.getPayload().getPullRequest();
        Comment comment = eventsModel.getPayload().getComment();
        spannableBuilder.bold("reviewed")
                .append(" ")
                .bold("pull request")
                .append(" ")
                .bold("in")
                .append(" ")
                .append(eventsModel.getRepo().getName())
                .bold("#")
                .bold(String.valueOf(pullRequest.getNumber()));
        if (comment.getBody() != null) {
            MarkDownProvider.stripMdText(description, comment.getBody().replaceAll("\\r?\\n|\\r", " "));
            description.setVisibility(View.VISIBLE);
        } else {
            description.setText("");
            description.setVisibility(View.GONE);
        }
    }

    private void appendPullRequestEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        PullRequest issue = eventsModel.getPayload().getPullRequest();
        String action = eventsModel.getPayload().getAction();
        if ("synchronize".equals(action)) {
            action = "updated";
        }
        if (eventsModel.getPayload().getPullRequest().isMerged()) {
            action = "merged";
        }
        spannableBuilder.bold(action)
                .append(" ")
                .bold("pull request")
                .append(" ")
                .append(eventsModel.getRepo().getName())
                .bold("#")
                .bold(String.valueOf(issue.getNumber()));
        if ("opened".equals(action) || "closed".equals(action)) {
            if (issue.getTitle() != null) {
                MarkDownProvider.stripMdText(description, issue.getTitle().replaceAll("\\r?\\n|\\r", " "));
                description.setVisibility(View.VISIBLE);
            } else {
                description.setText("");
                description.setVisibility(View.GONE);
            }
        }
    }

    private void appendPublicEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        String action = "public";
        if (eventsModel.getPayload() != null && "privatized".equalsIgnoreCase(eventsModel.getPayload().getAction())) {
            action = "private";
        }
        spannableBuilder.append("made")
                .append(" ")
                .append(eventsModel.getRepo().getName())
                .append(" ")
                .append(action);
    }

    private void appendMemberEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        User user = eventsModel.getPayload().getMember();
        spannableBuilder.bold("added")
                .append(" ")
                .append(user != null ? user.getLogin() + " " : "")
                .append("as a collaborator")
                .append(" ")
                .append("to")
                .append(" ")
                .append(eventsModel.getRepo().getName());
    }

    private void appendIssueEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        Issue issue = eventsModel.getPayload().getIssue();
        boolean isLabel = "label".equals(eventsModel.getPayload().getAction());
        LabelModel label = isLabel ? issue.getLabels() != null && !issue.getLabels().isEmpty()
                                     ? issue.getLabels().get(issue.getLabels().size() - 1) : null : null;
        spannableBuilder.bold(isLabel && label != null ? ("Labeled " + label.getName()) : eventsModel.getPayload().getAction())
                .append(" ")
                .bold("issue")
                .append(" ")
                .append(eventsModel.getRepo().getName())
                .bold("#")
                .bold(String.valueOf(issue.getNumber()));
        if (issue.getTitle() != null) {
            MarkDownProvider.stripMdText(description, issue.getTitle().replaceAll("\\r?\\n|\\r", " "));
            description.setVisibility(View.VISIBLE);
        } else {
            description.setText("");
            description.setVisibility(View.GONE);
        }
    }

    private void appendIssueCommentEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        Comment comment = eventsModel.getPayload().getComment();
        Issue issue = eventsModel.getPayload().getIssue();
        spannableBuilder.bold("commented")
                .append(" ")
                .bold("on")
                .append(" ")
                .bold(issue.getPullRequest() != null ? "pull request" : "issue")
                .append(" ")
                .append(eventsModel.getRepo().getName())
                .bold("#")
                .bold(String.valueOf(issue.getNumber()));
        if (comment.getBody() != null) {
            MarkDownProvider.stripMdText(description, comment.getBody().replaceAll("\\r?\\n|\\r", " "));
            description.setVisibility(View.VISIBLE);
        } else {
            description.setText("");
            description.setVisibility(View.GONE);
        }
    }

    private void appendGollumEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        List<WikiModel> wiki = eventsModel.getPayload().getPages();
        if (wiki != null && !wiki.isEmpty()) {
            for (WikiModel wikiModel : wiki) {
                spannableBuilder.bold(wikiModel.getAction())
                        .append(" ")
                        .append(wikiModel.getPageName())
                        .append(" ");
            }
        } else {
            spannableBuilder.bold(resources.getString(R.string.gollum))
                    .append(" ");
        }
        spannableBuilder
                .append(eventsModel.getRepo().getName());

    }

    private void appendGistEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        String action = eventsModel.getPayload().getAction();
        action = "create".equals(action) ? "created" : "update".equals(action) ? "updated" : action;
        spannableBuilder.bold(action)
                .append(" ")
                .append(itemView.getResources().getString(R.string.gist))
                .append(" ")
                .append(eventsModel.getPayload().getGist().getGistId());
    }

    private void appendForkEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        spannableBuilder.bold("forked")
                .append(" ")
                .append(eventsModel.getRepo().getName());
    }

    private void appendFollowEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        spannableBuilder.bold("started following")
                .append(" ")
                .bold(eventsModel.getPayload().getTarget().getLogin());
    }

    private void appendDownloadEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        spannableBuilder.bold("uploaded a file")
                .append(" ")
                .append(eventsModel.getPayload().getDownload() != null ? eventsModel.getPayload().getDownload().getName() : "")
                .append(" ")
                .append("to")
                .append(" ")
                .append(eventsModel.getRepo().getName());
    }

    private void appendCreateEvent(SpannableBuilder spannableBuilder, Event eventsModel) {
        PayloadModel payloadModel = eventsModel.getPayload();
        String refType = payloadModel.getRefType();
        spannableBuilder
                .bold("created")
                .append(" ")
                .append(refType)
                .append(" ")
                .append(!"repository".equalsIgnoreCase(refType) ? payloadModel.getRef() + " " : "")
                .bold("at")
                .append(" ")
                .append(eventsModel.getRepo().getName());
        if (payloadModel.getDescription() != null) {
            MarkDownProvider.stripMdText(description, payloadModel.getDescription().replaceAll("\\r?\\n|\\r", " "));
            description.setVisibility(View.VISIBLE);
        } else {
            description.setText("");
            description.setVisibility(View.GONE);
        }
    }

    private void appendWatch(SpannableBuilder spannableBuilder, EventsType type, Event eventsModel) {
        spannableBuilder.bold(resources.getString(type.getType()).toLowerCase())
                .append(" ")
                .append(eventsModel.getRepo().getName());
    }

    private void appendCommitComment(SpannableBuilder spannableBuilder, Event eventsModel) {
        Comment comment = eventsModel.getPayload().getCommitComment() == null ? eventsModel.getPayload().getComment() : eventsModel.getPayload()
                .getCommitComment();
        String commitId = comment != null && comment.getCommitId() != null && comment.getCommitId().length() > 10 ?
                          comment.getCommitId().substring(0, 10) : null;
        spannableBuilder.bold("commented")
                .append(" ")
                .bold("on")
                .append(" ")
                .bold("commit")
                .append(" ")
                .append(eventsModel.getRepo().getName())
                .url(commitId != null ? "@" + commitId : "");
        if (comment != null && comment.getBody() != null) {
            MarkDownProvider.stripMdText(description, comment.getBody().replaceAll("\\r?\\n|\\r", " "));
            description.setVisibility(View.VISIBLE);
        } else {
            description.setText("");
            description.setVisibility(View.GONE);
        }
    }

    private void appendActor(@NonNull Event eventsModel, SpannableBuilder spannableBuilder) {
        if (eventsModel.getActor() != null) {
            spannableBuilder.append(eventsModel.getActor().getLogin()).append(" ");
        }
    }

    private void appendAvatar(@NonNull Event eventsModel) {
        if (avatar != null) {
            if (eventsModel.getActor() != null) {
                avatar.setUrl(eventsModel.getActor().getAvatarUrl(), eventsModel.getActor().getLogin(),
                        eventsModel.getActor().isOrganizationType(),
                        LinkParserHelper.isEnterprise(eventsModel.getActor().getHtmlUrl()));
            } else {
                avatar.setUrl(null, null, false, false);
            }
        }
    }
}

File: app/src/main/java/com/fastaccess/github/ui/adapter/viewholder/FeedsViewHolder.kt
package com.fastaccess.github.ui.adapter.viewholder

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import com.fastaccess.data.persistence.models.FeedModel
import com.fastaccess.domain.response.enums.EventsType
import com.fastaccess.github.R
import com.fastaccess.github.extensions.replaceAllNewLines
import com.fastaccess.github.extensions.timeAgo
import com.fastaccess.github.ui.adapter.base.BaseViewHolder
import com.fastaccess.markdown.MarkdownProvider
import com.fastaccess.markdown.widget.SpannableBuilder
import kotlinx.android.synthetic.main.feeds_row_item.view.*

/**
 * Created by Kosh on 20.10.18.
 */
class FeedsViewHolder(
    parent: ViewGroup
) : BaseViewHolder<FeedModel>(LayoutInflater.from(parent.context)
    .inflate(R.layout.feeds_row_item, parent, false)) {

    override fun bind(item: FeedModel) {
        itemView.apply {
            feedDescription.text = ""
            feedDescription.isVisible = false
            feedTitle.text = context.getString(item.type?.titleId ?: 0).toLowerCase()
            when (item.type) {
                EventsType.WatchEvent -> watchEvent(this, item)
                EventsType.CreateEvent -> createEvent(this, item)
                EventsType.CommitCommentEvent -> commitCommentEvent(this, item)
                EventsType.DownloadEvent -> downloadEvent(this, item)
                EventsType.FollowEvent -> followEvent(this, item)
                EventsType.ForkEvent -> forkEvent(this, item)
                EventsType.GistEvent -> gistEvent(this, item)
                EventsType.GollumEvent -> gollumEvent(this, item)
                EventsType.IssueCommentEvent -> issueCommentEvent(this, item)
                EventsType.IssuesEvent -> issueEvent(this, item)
                EventsType.MemberEvent -> memberEvent(this, item)
                EventsType.PublicEvent -> publicEvent(this, item)
                EventsType.PullRequestEvent -> pullRequestEvent(this, item)
                EventsType.PullRequestReviewCommentEvent -> pullRequestReviewCommentEvent(this, item)
                EventsType.PullRequestReviewEvent -> pullRequestReviewEvent(this, item)
                EventsType.RepositoryEvent -> repositoryEvent(this, item)
                EventsType.PushEvent -> pushEvent(this, item)
                EventsType.TeamAddEvent -> teamAddedEvent(this, item)
                EventsType.DeleteEvent -> deleteEvent(this, item)
                EventsType.ReleaseEvent -> releaseEvent(this, item)
                EventsType.ForkApplyEvent -> forkApplyEvent(this, item)
                EventsType.OrgBlockEvent -> orgBlockEvent(this, item)
                EventsType.ProjectCardEvent -> projectCardEvent(this, item)
                EventsType.ProjectColumnEvent -> projectColumnEvent(this, item)
                EventsType.OrganizationEvent -> organizationEvent(this, item)
                EventsType.ProjectEvent -> projectEvent(this, item)
                else -> otherEvent(this, item)
            }
            dateWithIcon.text = item.createdAt?.timeAgo()
            stateIcon.setImageResource(item.type?.drawableRes ?: 0)
            userIcon.loadAvatar(item.actor?.avatarUrl, item.actor?.url)
        }
    }

    private fun otherEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold(item.payload?.action)
            .space()
            .append(item.repo?.name)
    }

    private fun organizationEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold(item.payload?.action?.replace("_", ""))
            .space()
            .append(item.payload?.invitation?.login)
            .space()
            .append(item.payload?.organization?.login)
    }

    private fun projectEvent(view: View, item: FeedModel) = projectCardEvent(view, item, false)

    private fun projectColumnEvent(view: View, item: FeedModel) = projectCardEvent(view, item, true)

    private fun projectCardEvent(view: View, item: FeedModel, isColumn: Boolean = false) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold(item.payload?.action)
            .space()
            .append(if (isColumn) "column" else "project")
            .space()
            .append(item.payload?.organization?.login)
    }

    private fun orgBlockEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold(item.payload?.action)
            .space()
            .append(item.payload?.blockedUser?.login)
            .space()
            .append(item.payload?.organization?.login)
    }

    private fun forkApplyEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold(item.payload?.head)
            .space()
            .append(item.payload?.before)
            .space()
            .append(item.repo?.name)
    }

    private fun releaseEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("released")
            .space()
            .append(item.payload?.release?.name)
            .space()
            .append(item.repo?.name)

    }

    private fun deleteEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("deleted")
            .space()
            .append(item.payload?.refType)
            .space()
            .bold("at ")
            .append(item.repo?.name)
    }

    private fun teamAddedEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("added")
            .space()
            .append(item.payload?.user?.login ?: item.repo?.name)
            .space()
            .bold("in ")
            .append(item.payload?.team?.name ?: item.payload?.team?.slug)

    }

    private fun pushEvent(view: View, item: FeedModel) {
        val ref = if (item.payload?.ref?.startsWith("refs/heads/") == true) {
            item.payload?.ref?.substring(11)
        } else {
            item.payload?.ref
        }
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("push to")
            .space()
            .append(ref)
            .space()
            .bold("at")
            .space()
            .append(item.repo?.name)

        val builder = SpannableBuilder.builder()
        val commits = item.payload?.commits
        if (commits?.isNotEmpty() == true) {
            if (commits.size != 1) {
                builder.append("${item.payload?.size}")
                    .bold(" new commits")
                    .newline()
            } else {
                builder.bold("1 new commit").newline()
            }
            commits.take(5)
                .filter { !it.sha.isNullOrEmpty() }
                .forEach {
                    val sha = if ((it.sha?.length ?: 0) > 7) {
                        it.sha?.subSequence(0, 7)
                    } else {
                        it.sha
                    }
                    builder.url(sha ?: "")
                        .space()
                        .append(it.message?.replaceAllNewLines())
                        .newline()
                }
            view.feedDescription.maxLines = 5
            view.feedDescription.text = builder
            view.feedDescription.isVisible = true
        } else {
            view.feedDescription.maxLines = 2
        }
    }

    private fun repositoryEvent(view: View, item: FeedModel) = publicEvent(view, item)

    private fun pullRequestReviewEvent(view: View, item: FeedModel) = pullRequestReviewCommentEvent(view, item)

    private fun pullRequestReviewCommentEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold(if (item.payload?.comment != null) "commented on a review in " else "reviewed a pull request in")
            .space()
            .append("${item.repo?.name}")
            .bold("#${item.payload?.pullRequest?.number}")
        view.feedDescription.apply {
            text = item.payload?.comment?.body ?: ""
            isVisible = !item.payload?.comment?.body.isNullOrEmpty()
        }
    }

    private fun pullRequestEvent(view: View, item: FeedModel) {
        val pullRequest = item.payload?.pullRequest
        val action = when {
            "synchronize" == item.payload?.action -> "updated"
            pullRequest?.isMerged == true -> "merged"
            else -> item.payload?.action
        }

        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold(action ?: "")
            .space()
            .append("${item.repo?.name}")
            .bold("#${pullRequest?.number}")
        if ("opened" == action || "closed" == action) {
            view.feedDescription.text = pullRequest?.title?.replaceAllNewLines() ?: ""
            view.feedDescription.isVisible = !pullRequest?.title?.replaceAllNewLines().isNullOrEmpty()
        }
    }

    private fun publicEvent(view: View, item: FeedModel) {
        val action = if ("privatized" == item.payload?.action) {
            "private"
        } else {
            "public"
        }
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .append(item.repo?.name)
            .space()
            .bold(action)
    }

    private fun memberEvent(view: View, item: FeedModel) {
        val user = item.payload?.member
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("added")
            .space()
            .append(user?.login)
            .space()
            .append("as a collaborator to ")
            .append(item.repo?.name)
    }

    private fun issueEvent(view: View, item: FeedModel) {
        val issue = item.payload?.issue
        val label = item.payload?.issue?.labels?.lastOrNull()
        val isLabel = "label" == item.payload?.action
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold(if (isLabel && label != null) "Labeled ${label.name}" else item.payload?.action ?: "")
            .space()
            .append(item.repo?.name)
            .space()
            .bold("#${issue?.number}")
        view.feedDescription.text = issue?.title?.replaceAllNewLines() ?: ""
        view.feedDescription.isVisible = !issue?.title?.replaceAllNewLines().isNullOrEmpty()
    }

    private fun issueCommentEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("commented on ${if (item.payload?.issue?.htmlUrl?.contains("/pull/", true) == true) "a pull request" else "an issue"}")
            .space()
            .append("${item.repo?.name}")
            .bold("#${item.payload?.issue?.number}")

        view.feedDescription.text = item.payload?.comment?.body?.replaceAllNewLines()?.let { MarkdownProvider.stripMd(it) }
        view.feedDescription.isVisible = !item.payload?.comment?.body?.replaceAllNewLines().isNullOrEmpty()
    }

    private fun gollumEvent(view: View, item: FeedModel) {
        val wikies = item.payload?.pages
        val builder = SpannableBuilder.builder().append(item.actor?.login)
            .space()
        if (wikies?.isNotEmpty() == true) {
            wikies.forEach {
                builder.bold(it.action)
                    .space()
                    .append(it.packageName)
                    .space()
            }
        } else {
            builder.bold(view.resources.getString(R.string.gollum))
                .space()
        }
        builder.append(item.repo?.name)
        view.feedTitle.text = builder
    }

    private fun gistEvent(view: View, item: FeedModel) {
        val action = if ("create" == item.payload?.action) "created" else if ("update" == item.payload?.action) "updated" else item.payload?.action

        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("$action ${view.resources.getString(R.string.gist).toLowerCase()}")
            .space()
            .append(item.payload?.gist?.id)
    }

    private fun forkEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("forked")
            .space()
            .append(item.repo?.name)
    }

    private fun followEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("started following")
            .space()
            .append(item.payload?.target?.login)
    }

    private fun downloadEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("uploaded a file")
            .space()
            .append("${item.payload?.download?.name}")
            .space()
            .append("to")
            .space()
            .bold(item.repo?.name ?: "")
    }

    private fun commitCommentEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold("commented on commit")
            .space()
            .append("${item.repo?.name}")
            .bold("#${item.payload?.issue?.number?.toString()}")
    }

    private fun watchEvent(view: View, item: FeedModel) {
        view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
            .space()
            .bold(view.context.getString(item.type?.titleId ?: 0).toLowerCase())
            .space()
            .append(item.repo?.name)
    }

    private fun createEvent(view: View, item: FeedModel) {
        item.payload?.let {
            view.feedTitle.text = SpannableBuilder.builder().append(item.actor?.login)
                .space()
                .bold("created")
                .space()
                .append(it.refType)
                .space()
                .append(if ("repository" != it.refType) it.ref else "")
                .space()
                .append("at")
                .space()
                .append(item.repo?.name)
            view.feedDescription.text = it.description?.replaceAllNewLines()?.let { MarkdownProvider.stripMd(it) }
            view.feedDescription.isVisible = !it.description?.replaceAllNewLines().isNullOrEmpty()
        }
    }
}